includes/class-sensei-course-access.php

```diff
@@ -0,0 +1,294 @@
+	 * @return bool
+	 */
+	private function get_default_access( $user_id ) {
+		// No access provider has explicitly granted or restricted access to the user. Use Sensei's default.
+		if ( Sensei_Utils::has_started_course( $this->course_id, $user_id ) ) {
+			return true;
+		}
+
+		return ! Sensei()->settings->get( 'access_permission' ) || sensei_all_access();
+	}
+
+	/**
+	 * Get the course log meta key.
+	 *
+	 * @return string
+	 */
+	private function get_course_log_meta_key() {
+		return self::META_PREFIX_COURSE_LOG . $this->course_id;
+	}
+
+	/**
+	 * Get an array of all the instantiated course access providers.
+	 *
+	 * @return Sensei_Course_Access_Provider_Interface[]
+	 */
+	private static function get_course_access_providers() {
+		if ( ! isset( self::$access_providers ) ) {
+			self::$access_providers = [];
+
+			/**
+			 * Fetch all registered course access providers.
+			 *
+			 * @since 2.x.x
+			 *
+			 * @param string[] $provider_classes List of access providers classes.
+			 */
+			$provider_classes = apply_filters( 'sensei_course_access_providers', [] );
+			foreach ( $provider_classes as $provider_class ) {
+				if ( ! class_exists( $provider_class ) || ! is_a( $provider_class, 'Sensei_Course_Access_Provider_Interface', true ) ) {
+					continue;
+				}
+
+				self::$access_providers[ $provider_class::get_id() ] = new $provider_class();
+			}
+		}
+
+		return self::$access_providers;
+	}
+
+	/**
+	 * Get the hash of the current versions of all course access providers.
+	 *
+	 * @return string
+	 */
+	public function get_course_access_providers_version() {
+		if ( ! isset( self::$access_providers_version ) ) {
+			$access_providers               = self::get_course_access_providers();
+			self::$access_providers_version = self::hash_course_access_provider_versions( array_keys( $access_providers ) );
+		}
+
+		return self::$access_providers_version;
+	}
+
+	/**
+	 * Generates a hash of all the access provider versions.
+	 *
+	 * @param string[] $access_providers Array of access provider class names.
+	 * @return string
+	 */
+	public static function hash_course_access_provider_versions( $access_providers ) {
+		$versions = [];
+		foreach ( $access_providers as $access_provider_class ) {
+			if ( ! is_a( $access_provider_class, 'Sensei_Course_Access_Provider_Interface', true ) ) {
+				continue;
+			}
+
+			$versions[ $access_provider_class ] = $access_provider_class::get_version();
+		}
+
+		return md5( wp_json_encode( $versions ) );
+	}
+
+	/**
+	 * Check if we should use the legacy access check. Legacy access check
+	 * uses course enrollment to determine access.
+	 *
+	 * return bool
+	 */
+	public static function use_legacy_access_check() {
+		$use_legacy_access_check = false;
+
+		// Check if WCPC is around but not offering access providers (an old version).
+		if (
+			class_exists( '\Sensei_WC_Paid_Courses\Sensei_WC_Paid_Courses' ) &&
+			! class_exists( '\Sensei_WC_Paid_Courses\Course_Access_Providers' )
```

Comment:
I think we might need to do this WCPC check here, but not sure?

Reply:
I'm going to keep this for now, but probably replace it with a default course provider.